app/routes/content.js

```diff
@@ -48,31 +48,35 @@ export default Ember.Route.extend(Analytics, ResetScrollMixin, SetupSubmitContro
         return this._super(...arguments);
     },
     afterModel(preprint) {
-        // If we're already on a branded page, stay there.
-        if (!this.get('theme.isProvider')) {
-            preprint.get('provider')
-                .then(provider => {
-                    const providerId = provider.get('id');
-
-                    // If we're on the OSF and the provider is OSF, stay on OSF
-                    if (providerId === 'osf')
-                        return;
-
-                    // Otherwise, redirect to the branded page
-                    // Hard redirect instead of transition, in anticipation of Phase 2 where providers will have their own domains.
-                    const {origin, pathname, search} = window.location;
-
-                    const url = [
-                        origin,
-                        'preprints',
-                        providerId,
-                        `${pathname.replace(/^\/(preprints\/)?/, '')}${search}`
-                    ].join('/');
-
-                    window.history.replaceState({}, document.title, url);
-                    window.location.replace(url);
-                });
-        }
+        // Redirect if necessary
+        preprint.get('provider')
+            .then(provider => {
+                const providerId = provider.get('id');
+                const themeId = this.get('theme.id');
+                const isOSF = providerId === 'osf';
+
+                // If we're on the proper branded site, stay there.
+                if ((!themeId && isOSF) || themeId === providerId)
+                    return;
+
+                // Otherwise, redirect to the branded page
+                // Hard redirect instead of transition, in anticipation of Phase 2 where providers will have their own domains.
+                const {origin, search} = window.location;
+
+                const urlParts = [
+                    origin
+                ];
+
+                if (!isOSF)
+                    urlParts.push('preprints', providerId);
+
+                urlParts.push(`${preprint.get('id')}${search}`);
```

Comment:
@binoculars I would put a slash in between the preprint id and the search item.  Otherwise I am seeing `/preprints/socarxiv/m86jw/?edit` redirect to `/m86jw?edit` redirect to `m86jw/?edit`.
```js
urlParts.push(`${preprint.get('id')}/${search}`);
```